src/common/packager.ts

```diff
@@ -41,77 +51,64 @@ export class Packager {
-                                args = args.concat("--resetCache");
-                            }
-                            let reactEnv = Object.assign({}, process.env, {
-                                REACT_DEBUGGER: "echo A debugger is not needed: ",
-                                REACT_EDITOR: this.openFileAtLocationCommand(),
-                            });
-
-                            Log.logMessage("Starting Packager");
-                            // The packager will continue running while we debug the application, so we can"t
-                            // wait for this command to finish
-
-                            let spawnOptions = { env: reactEnv };
-
-                            const packagerSpawnResult = new CommandExecutor(this.projectPath).spawnReactPackager(args, spawnOptions);
-                            this.packagerProcess = packagerSpawnResult.spawnedProcess;
-                            packagerSpawnResult.outcome.done(() => {}, () => {}); /* Q prints a warning if we don't call .done().
-                                                                                     We ignore all outcome errors */
-                            return packagerSpawnResult.startup;
-                        });
+                if (running && this.packagerRunningAs === PackagerRunAs.REACT_NATIVE) {
+                    return this.killPackagerProcess()
+                        .then(() =>
+                            this.start(port, PackagerRunAs.EXPONENT));
+                } else if (running && this.packagerRunningAs === PackagerRunAs.NOT_RUNNING) {
+                    Log.logWarning(ErrorHelper.getWarning("Packager running outside of VS Code. To avoid issues with exponent make sure it is running with .vscode/ as a root."));
+                    return Q.resolve<void>(void 0);
+                } else if (this.packagerRunningAs !== PackagerRunAs.EXPONENT) {
+                    return this.start(port, PackagerRunAs.EXPONENT);
                 }
             })
             .then(() =>
-                this.awaitStart())
-            .then(() => {
-                if (executedStartPackagerCmd) {
-                    Log.logMessage("Packager started.");
-                } else {
-                    Log.logMessage("Packager is already running.");
-                    if (!this.packagerProcess) {
-                        Log.logWarning(ErrorHelper.getWarning("React Native Packager running outside of VS Code. If you want to debug please use the 'Attach to packager' option"));
-                    }
-                }
+                XDL.setOptions(this.projectPath, { packagerPort: port })
+            ).then(() =>
+                XDL.startExponentServer(this.projectPath)
+            ).then(() =>
+                XDL.startTunnels(this.projectPath)
+            ).then(() =>
+                XDL.getUrl(this.projectPath, { dev: true, minify: false })
+            ).then(exponentUrl => {
```

Comment:
Each of these `XDL.*` functions return a promise, and we want them to execute in the correct order so we need to wait for each step to complete before beginning the next.